idseq_dag/util/m8.py

```diff
@@ -181,21 +183,49 @@ def accumulate(hits, accession_id):
                 # provided by other accessions. This occurs a lot and
                 # handling it in this way seems to work well.
                 continue
-            hits[level][taxid_at_level] = accession_id
+            accession_list = hits[level].get(taxid_at_level, []) + [accession_id]
+            hits[level][taxid_at_level] = accession_list
         return blacklisted
 
+    def most_frequent_accession(accession_list):
+        counts = Counter(accession_list)
+        return counts.most_common(1)[0][0]
+
+
     def call_hit_level(hits, blacklisted = False):
-        ''' Call hit if read not blacklisted and only one taxid at level '''
+        ''' TO BE DEPRECATED: Call hit if read not blacklisted and only one taxid at level '''
         if not blacklisted:
             for level, hits_at_level in enumerate(hits):
                 if len(hits_at_level) == 1:
-                    taxid, accession_id = hits_at_level.popitem()
+                    taxid, accession_list = hits_at_level.popitem()
+                    accession_id = most_frequent_accession(accession_list)
                     return level + 1, taxid, accession_id
         return -1, "-1", None
 
+    def call_hit_level_v2(hits, blacklisted = False):
+        ''' Always call hit at the species level with the taxid with most matches '''
+        if not blacklisted:
+            species_level_hits = hits[0]
+            max_match = 0
+            taxid_candidates = []
+            for taxid, accession_list in species_level_hits.items():
+                accession_len = len(accession_list)
+                if accession_len > max_match:
+                    taxid_candidates = [taxid]
+                    max_match = accession_len
+                elif accession_len == max_match:
+                    taxid_candidates.append(taxid)
+            if max_match > 0:
+                selected_taxid = taxid_candidates[0]
+                if len(taxid_candidates) > 1:
+                    selected_taxid = random.sample(taxid_candidates, 1)[0]
```

Comment:
This is the place where in hit calling v1 we would toss the read into a non-species-specific bucket, but now with this v2 we are choosing a random species, just like we did implicitly before hit calling v1.   @cdebourcy 

My question is, why do we go through all this trouble, if blast is going to determine a more accurate assignment later?   Is it just to have temporary results to look at, while blast is crunching?   And we've chosen to recreate the pre-hit-calling results as those temporary results --- instead of the hit-calling results?   Why?